S2A handshaker client, part 2

matthewstevenson88 opened the pull request:
This PR is a follow-up to #21. It contains the implementation of the methods of the `s2a_handshaker_client` that are responsible for calling the S2A handshaker service.

matthewstevenson88 commented:
This PR is ready for review, once PR #21 is merged. The implementation and the tests are nearly identical to the ALTS counterparts, so it should be pretty easy to review.

matthewstevenson88 commented:
This PR is ready for review!

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -86,11 +86,14 @@ tsi_result S2AHandshakerClient::ClientStart() {
+    }
   }
-  return call_result;
+  return TSI_OK;
```
Is it possible to refactor the code from lines 89 to 94? It seems to be used repeatedly in three different places.

matthewstevenson88 replied:
Done in next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -175,18 +181,34 @@ tsi_result S2AHandshakerClient::Next(grpc_slice* bytes_received) {
+    gpr_free(status_details);
+  }
+  client->MaybeCompleteTsiNext(/* receive_status_finished=*/true,
+                               /* pending_recv_message_result=*/nullptr);
```
Don't we also need a global queue (similar to ALTS) to handle a large number of concurrent connection requests?

matthewstevenson88 replied:
As we discussed offline, we do not need the global queue. Limiting the number of concurrent connection requests will be handled by the S2A service.

yihuazhang replied:
Ack.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.h:
```diff
@@ -111,16 +111,43 @@ class S2AHandshakerClient {
+      bool receive_status_finished,
+      s2a_recv_message_result* pending_recv_message_result);
+
+  /** This method dereferences all of the data owned by this S2A handshaker
```
The comment is confusing a little bit. I think it dereferences a handshaker client and clean up all of its fields once the ref count is zero.

matthewstevenson88 replied:
I've made this comment more precise in the next commit. PTAL.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -235,15 +259,15 @@ tsi_result s2a_handshaker_client_create(
     const grpc_s2a_credentials_options* options, const grpc_slice& target_name,
     grpc_iomgr_cb_func grpc_cb, tsi_handshaker_on_next_done_cb cb,
-    void* user_data, bool is_client, S2AHandshakerClient** client) {
+    void* user_data, bool is_client, bool is_test, S2AHandshakerClient** client) {
```
Please also change the function name to adhere to C++ naming convention.

matthewstevenson88 replied:
Done in next commit (and same for the destroy method).

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -259,14 +283,16 @@ static void S2AHandshakerCallUnref(void* arg, grpc_error* error) {
                               GRPC_ERROR_NONE);
     }
+    gpr_free(refs_);
+    refs_ = nullptr;
```
Why do we set it `nullptr` after freeing it?

matthewstevenson88 replied:
It is not needed. Good catch! I've removed it in the next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -275,11 +301,14 @@ S2AHandshakerClient::~S2AHandshakerClient() {
     grpc_slice_unref_internal(recv_bytes_);
     grpc_slice_unref_internal(target_name_);
     gpr_free(buffer_);
+    buffer_ = nullptr;
```
Same as above.

matthewstevenson88 replied:
Also removed it.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.h:
```diff
@@ -186,10 +213,17 @@ class S2AHandshakerClient {
+  s2a_recv_message_result* pending_recv_message_result_ = nullptr;
+  /** This variable should be set to true iff the S2A handshaker client instance
+   *  is instantiated as part of a test. **/
+  bool is_test_ = false;
```
Can we consolidate these two variables? `is_test_` and `no_calls_for_testing_`

matthewstevenson88 replied:
I kept them separate for the following reason: there are basic unit tests in `s2a_handshaker_client_test.cc`, and then unit tests for the methods that call the S2A service in `s2a_handshaker_client_caller_test.cc`. 

The variable `is_test_` is set to `true` in both cases (and in certain TSI handshaker tests). The variable `no_calls_for_testing_` is set to `true` in the former case, and `false` in the latter case.

Note that only `is_test_` must be specified in the constructor of the `S2AHandshakerClient`.

What are your thoughts on this strategy?

yihuazhang replied:
IIUC, `MakeGrpcCall` is not actually called in both cases since `grpc_caller_` callback (which talks to the handshaker server) is not called so not sure why we treat those two tests differently.

yihuazhang replied:
I did not see this comment is addressed.

matthewstevenson88 replied:
Sorry Yihua I missed this comment on my last pass. The issue is the following: for the tests in `s2a_handshaker_client_test.cc`, there is no `grpc_caller_` set and, if the `no_calls_for_testing_` variable is not present in `MakeGrpcCallUtil, then `MakeGrpcCall` (and hence `grpc_caller_`) is called. This leads to an error in the test.

After looking at this more carefully, I think it makes sense to do the following:
1. Remove the `s2a_handshaker_client_test.cc` file entirely and add any edge cases that it tests to `s2a_handshaker_client_caller_test.cc`.
2. Remove the `no_calls_for_testing_` variable.

WDYT?

yihuazhang replied:
The plan (removing the boolean flag and combining these two tests) SGTM.

matthewstevenson88 replied:
Sounds good. Made these changes in the next commit.

yihuazhang commented:
Sorry for the delay Matt. 

My major concern is I am not convinced about the correctness of some changes in this PR without seeing more comprehensive tests:   

1: Apolcyn@ made a series of changes to ALTS code to make it resistant against a large number of concurrent connection requests (https://github.com/grpc/grpc/pull/20722, https://github.com/grpc/grpc/pull/21189, https://github.com/grpc/grpc/pull/20929). In this PR, I noticed a similar but a little different changes are made (e.g., I did not see the use of connection queues) so not sure if the changes indeed work without doing more integration tests (similar to `alts_concurrent_connectivity_test.cc`). Basically, we did not verify any interaction to the handshake server. 

2: Do you think it is necessary to do some code refactoring between ALTS and S2A? If the long term plan is to move the S2A code out of gRPC repo, it may not be necessary though.

matthewstevenson88 commented:
Thanks for the comments @yihuazhang! I really appreciate them. I've addressed them all in the most recent commit, or responded to them directly.

Below, I will respond to your 2 comments above. (Some of this is repetition from our offline discussion.)

1. As we discussed offline, our team has agreed that the S2A service will be responsible for capping the number of concurrent connection requests. As a consequence, the global queue is not necessary. (There is an email chain with apolcyn@, on which you are cc-ed, that verifies what portions of code will be removed from ALTS along with the global queue. I've removed them in the S2A client already.)

As for the additional tests, I propose the following: when two more PR's are merged (this one and one other yet to be posted), there will be a complete TSI handshaker and handshaker client implementation. Once this is done, I will post a PR that contains the following additional tests:

- The S2A version of the fake handshaker service, with a cap on the number of concurrent connection requests.
- The S2A version of the tests in `alts_tsi_handshaker_test.cc` (i.e. the ones that use the fake handshaker service to mock a handshake).
- The S2A version of the `alts_concurrent_connectivity_test.cc`.

How does that plan look to you?

2. What code refactoring are you referring to? Once the S2A client is ultimately merged somewhere, I will do some reorganization of the ALTS code, namely putting some stuff in a shared library under `src/core/tsi`. Is this what you are referring to?

In terms of the long term plan, if we decide to move the S2A client out of gRPC, there will also be refactoring to avoid the use of gRPC-internal API's.

yihuazhang commented:
The plan SGTM. Regarding refactoring since it seems S2A will not be placed in OSS gRPC repo, let's do not do any refactoring regarding code sharing between ALTS and S2A.

matthewstevenson88 commented:
Per our offline discussion, we will create an S2A-dedicated copy of the `gsec_aead_crypter`, to avoid any refactoring of the ALTS code.

Granted the plan, is this PR good to go? Or did you want to take another look at it?

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.h:
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
-  void MaybeCompleteTsiNext(
-      bool receive_status_finished,
-      s2a_recv_message_result* pending_recv_message_result);
+  tsi_result MakeGrpcCallUtil(bool is_start);
```
Add some comments for it?

matthewstevenson88 replied:
Done in next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.h:
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
+  }
+
+  /** These methods are exposed for testing purposes only. If |is_test_| is
+   *  false, then these methods do nothing or return nullptr. **/
```
`and`?

matthewstevenson88 replied:
Sorry, I meant: if the return type is `void`, do nothing; if the return type is a pointer, then do nothing and return `nullptr`. I've clarified this comment in the next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc:
```diff
@@ -289,8 +305,49 @@ void s2a_handshaker_client_destroy(S2AHandshakerClient* client) {
+  }
+}
+
+grpc_metadata_array* S2AHandshakerClient::initial_metadata_for_testing() {
```
The logic here is a little confusing. I expected that if `is_test_` is true, `nullptr` should be returned which does not seem to be the case.

matthewstevenson88 replied:
Since this is a `for_testing` method, it should only return something interesting if `is_test_` is true. If `is_test_` is false, then it should return `nullptr`. (The confusion here might be related to the previous comment, about the comments on the `for_testing` functions that were unclear.)

yihuazhang replied:
I see. Please also add above comments in the code to make it clear.

matthewstevenson88 replied:
Clarified this comment in the next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc:
```diff
@@ -30,16 +32,208 @@ namespace experimental {
+  return TSI_OK;
 }
 
+/** The implementation of this method is nearly identical to its ALTS
```
Not sure if we want to keep this comment especially if we do not want to place the code in gRPC repo.

matthewstevenson88 replied:
Removed in next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc:
```diff
@@ -30,16 +32,208 @@ namespace experimental {
 
+const size_t kHandshakerClientOpNum = 4;
+
+/** The implementation of |make_grpc_call| is nearly identical to that of its
```
Similar to the comment below.

matthewstevenson88 replied:
Removed in next commit.

yihuazhang commented on src/core/tsi/s2a/handshaker/s2a_handshaker_client.h:
```diff
@@ -111,16 +111,47 @@ class S2AHandshakerClient {
 
-  /** This method is exposed for testing purposes only. **/
-  grpc_byte_buffer* get_send_buffer_for_testing();
+  void MaybeCompleteTsiNext(
```
Can we add some comments for it? especially why it is `Maybe`?

matthewstevenson88 replied:
I've added some comments in the next commit.

yihuazhang replied:
I do not think the comment is accurate. 

If the handshaker client has a final result or failure, it checks if the status is received or not, and if it is received then it will invoke the TSI callback and if the status is not received, it will not invoke the callback. If the final result is not received, it will go ahead execute the callback.

matthewstevenson88 replied:
I've corrected this in the newest commit. Thanks! :)

yihuazhang commented:
Thanks for the changes Matt. My major concern is that since we do not want to place the code inside gRPC repo (either a short or long term), I think the comments in the code should not or at least minimize its reference to `ALTS`. Please add more comments to some of the API's (`MaybeCompleteTsiNext`) to make it easier to understand in future. You can find some conversations I had with Alex in https://github.com/grpc/grpc/pull/20929/files about it.

matthewstevenson88 commented:
Thanks for the comments @yihuazhang! I've responded to them all. In particular, I've removed all references to `ALTS` and added more comments (e.g. to the `MaybeCompleteTsiNext` API).

matthewstevenson88 commented:
Thanks for all the comments @yihuazhang! I've implemented the changes we discussed in the most recent commit.

matthewstevenson88 commented:
Thanks for the review @yihuazhang!

Squashing and merging now.
